clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3783,6 +3612,363 @@ static void handleSYCLIntelMaxWorkGroupSize(Sema &S, Decl *D,
+  const auto *R1WGSXDimExpr = dyn_cast<ConstantExpr>(R1WGSXDim);
+  const auto *R1WGSYDimExpr = dyn_cast<ConstantExpr>(R1WGSYDim);
+  const auto *R1WGSZDimExpr = dyn_cast<ConstantExpr>(R1WGSZDim);
+
+  if (!RWGSXDimExpr || !RWGSYDimExpr || !RWGSZDimExpr || !R1WGSXDimExpr ||
+      !R1WGSYDimExpr || !R1WGSZDimExpr)
+    return false;
+
+  // Otherwise, check if value of reqd_work_group_size argument is
+  // less than value of reqd_work_group_size attribute argument.
+  bool CheckFirstArgument = S.getLangOpts().OpenCL
+                                ? RWGSXDimExpr->getResultAsAPSInt() <
+                                      R1WGSZDimExpr->getResultAsAPSInt()
+                                : RWGSXDimExpr->getResultAsAPSInt() <
+                                      R1WGSXDimExpr->getResultAsAPSInt();
+
+  bool CheckSecondArgument =
+      RWGSYDimExpr->getResultAsAPSInt() < R1WGSYDimExpr->getResultAsAPSInt();
+
+  bool CheckThirdArgument = S.getLangOpts().OpenCL
+                                ? RWGSZDimExpr->getResultAsAPSInt() <
+                                      R1WGSXDimExpr->getResultAsAPSInt()
+                                : RWGSZDimExpr->getResultAsAPSInt() <
+                                      R1WGSZDimExpr->getResultAsAPSInt();
+
+  return CheckFirstArgument || CheckSecondArgument || CheckThirdArgument;
+}
+
+// If the reqd_work_group_size attribute is specified on a declaration
+// along with num_simd_work_items, the required work group size specified
+// by num_simd_work_items attribute must evenly divide the index that
+// increments fastest in the reqd_work_group_size attribute.
+//
+// The arguments to reqd_work_group_size are ordered based on which index
+// increments the fastest. In OpenCL, the first argument is the index that
+// increments the fastest, and in SYCL, the last argument is the index that
+// increments the fastest.
+static bool CheckWorkGroupSize(Sema &S, const Expr *NSWIValue,
+                               const Expr *RWGSXDim, const Expr *RWGSZDim) {
+  // If any of the operand is still value dependent, we can't test anything.
+  const auto *NSWIValueExpr = dyn_cast<ConstantExpr>(NSWIValue);
+  const auto *RWGSXDimExpr = dyn_cast<ConstantExpr>(RWGSXDim);
+  const auto *RWGSZDimExpr = dyn_cast<ConstantExpr>(RWGSZDim);
+
+  if (!NSWIValueExpr || !RWGSXDimExpr || !RWGSZDimExpr)
+    return false;
+
+  // Otherwise, check which argument increments the fastest
+  // in OpenCL vs SYCL mode.
+  unsigned WorkGroupSize =
+      S.getLangOpts().OpenCL
+          ? (RWGSXDimExpr->getResultAsAPSInt()).getZExtValue()
+          : (RWGSZDimExpr->getResultAsAPSInt()).getZExtValue();
+
+  // Check if the required work group size specified by num_simd_work_items
+  // attribute must evenly divide the index that increments fastest in the
+  // reqd_work_group_size attribute.
+  return WorkGroupSize % (NSWIValueExpr->getResultAsAPSInt()).getSExtValue() !=
```

Comment:
```suggestion
  return WorkGroupSize % NSWIValueExpr->getResultAsAPSInt().getSExtValue() !=
```
Why is this using the SExtValue instead of the ZExtValue?

Reply:
>>Why is this using the SExtValue instead of the ZExtValue?

This is an issue with current implementation of num_simd_work_items attribute. I have updated the attribute so that it can use ZExtValue instead of the SExtValue.